### PR TITLE
no-invalid-template-strings logic errors

### PR DESCRIPTION
**Pull request by aervin**
- [x] Addresses an existing issue: #[3128](https://github.com/palantir/tslint/issues/3128)
- [x] New feature, bugfix, or enhancement
  - [x] Includes tests
- [ ] Documentation update

#### Overview of change:
Addresses nodes w/ multiple template strings

### CODE REVIEW

File: src/rules/noInvalidTemplateStringsRule.ts [hunk 1]
```diff
@@ -50,24 +50,18 @@ function walk(ctx: Lint.WalkContext<void>) {
...
-            const unescapedText = node.getFullText();
-            const preceedingCharacter = unescapedText.substr(unescapedText.search(findTemplateString) - 1, 1);
-            if (isBackslash(preceedingCharacter)) {
-                return;
+        const text = node.getText(ctx.sourceFile);
+        const findTemplateStrings = /\$\{/g;
+        const nextInstance = () => instance = findTemplateStrings.exec(text);
+        let instance: RegExpExecArray | null = nextInstance();
+        while (instance !== null) {
+            const preceedingCharacters = text.substr(instance.index - 2, 2);
```
Comment (ajafff) [inline src/rules/noInvalidTemplateStringsRule.ts]: better count the preceding backslashes. If it's an odd number, the `$` is escaped with a `\`

consider: `"\\\${}"` is escaped, but `"\\\\${}"` is not

File: src/rules/noInvalidTemplateStringsRule.ts [hunk 2]
```diff
@@ -50,24 +50,18 @@ function walk(ctx: Lint.WalkContext<void>) {
...
+            if (!isBackslash(preceedingCharacters[0]) && isBackslash(preceedingCharacters[1])) {
```
Comment (ajafff) [inline src/rules/noInvalidTemplateStringsRule.ts]: could use a comment to explain what it does

File: src/rules/noInvalidTemplateStringsRule.ts (hunk 1 above)
Comment (aervin) [inline src/rules/noInvalidTemplateStringsRule.ts]: Should I rethink the while loop approach then? Do you have advice for solving this particular problem?

File: src/rules/noInvalidTemplateStringsRule.ts [hunk 3]
```diff
@@ -50,28 +50,20 @@ function walk(ctx: Lint.WalkContext<void>) {
...
-        if (index !== -1) {
-            /**
-             * Support for ignoring case: '\${template-expression}'
-             */
-            const unescapedText = node.getFullText();
-            const preceedingCharacter = unescapedText.substr(unescapedText.search(findTemplateString) - 1, 1);
-            if (isBackslash(preceedingCharacter)) {
-                return;
+        const text = node.getText(ctx.sourceFile);
+        const findTemplateStrings = /([\\]?)+(\$\{)/g;
```
Comment (ajafff) [inline src/rules/noInvalidTemplateStringsRule.ts]: that should probably be 
```js
/\\*\$\{/g
```

File: src/rules/noInvalidTemplateStringsRule.ts [hunk 4]
```diff
@@ -50,28 +50,20 @@ function walk(ctx: Lint.WalkContext<void>) {
...
+        let instance = findTemplateStrings.exec(text);
+        while (instance !== null) {
+            const matchLength = instance[0].length;
+            const backslashCount = matchLength - 2;
+            const instanceIsEscaped = (matchLength > 2) ? (backslashCount % 2) === 1 : false;
```
Comment (ajafff) [inline src/rules/noInvalidTemplateStringsRule.ts]:
```ts
const instanceIsEscaped = backslashCount % 2 === 1;
```

File: src/rules/noInvalidTemplateStringsRule.ts [hunk 5]
```diff
@@ -50,28 +50,20 @@ function walk(ctx: Lint.WalkContext<void>) {
...
+            if (instanceIsEscaped) {
+                instance = findTemplateStrings.exec(text);
+                continue;
             }
-
-            const textStart = node.getStart() + 1;
-            ctx.addFailureAt(textStart + index, 2, Rule.FAILURE_STRING);
+            const start = node.getStart() + (instance.index + backslashCount);
+            ctx.addFailureAt(start, 2, Rule.FAILURE_STRING);
```
Comment (ajafff) [inline src/rules/noInvalidTemplateStringsRule.ts]: You avoid code duplication if you wrap these 2 lines in an IfStatement and remove the if-block in line 60-62 completely
```ts
if (!instanceIsEscaped) {
    const start = ...;
    ctx.addFailureAt(...);
}
instance = findTemplateStrings.exec(text);
```

File: test/rules/no-invalid-template-strings/test.ts.lint [hunk 6]
```diff
@@ -1,9 +1,23 @@
...
+                                           ~~ [0]
+                                                                                  ~~ [0]
+                                                                                                                ~~ [0]
+
+new (Tab.mixins.Controlled('MemberComments',Tab.mixins.Templated('<div mbr-comment-tab=\\${j}></div>',Tab.mixins.AlwaysHasData())))
+                                                                                         ~~ [0]
+
+new (Tab.mixins.Templated('<div mbr-comment-tab=\\\\\${b}></div>',Tab.mixins.AlwaysHasData('\\\\${c}'))))
+                                                                                                ~~ [0]
 "One plus one is ${1 + 1}.";
```
Comment (ajafff) [inline test/rules/no-invalid-template-strings/test.ts.lint]: please add a test case with multiple interpolations, one escaped and one not escaped:
```ts
"\${a} = ${a}";
         ~~ [0]
```

**Comment by ajafff** [comment]
Thanks @aervin

**Pull request merged**